base/operators.jl

```diff
@@ -224,16 +224,44 @@ function promote_shape(a::Dims, b::Dims)
     return a
 end
 
+# The lengths of the given indices, lowering : to the appropriate size
+stagedfunction index_lengths(A::AbstractArray, I...)
+    N = length(I)
+    ex = Expr(:tuple)
+    for d=1:N
+        if I[d] <: Colon
+            push!(ex.args, d < N  ? (:(size(A, $d))) :
+                           d == 1 ? (:(length(A)))   : (:(trailingsize(A, $d))))
+        elseif I[d] <: Real
+            push!(ex.args, 1)
+        else
+            push!(ex.args, :(length(I[$d])))
+        end
+    end
+    ex
+end
+
 # shape of array to create for getindex() with indexes I
 # drop dimensions indexed with trailing scalars
-index_shape(I::Real...) = ()
-index_shape(i, I...) = tuple(length(i), index_shape(I...)...)
+stagedfunction index_shape(A::AbstractArray, I...)
```

Comment:
While I haven't thought it through in detail, it seems like it should be possible if you follow a pattern similar to `next`, e.g.,
```
index_shape_dim(A, dim, i::Colon, I...) = tuple(size(A, dim), index_shape_dim(A, dim+1, I...)...)
index_shape_dim(A, dim, i, I...) = tuple(length(i), index_shape_dim(A, dim+1, I...)...)
```